h2o-r/h2o-package/R/connection.R

```diff
@@ -163,6 +163,14 @@ h2o.init <- function(ip = "localhost", port = 54321, startH2O = TRUE, forceDL =
       stop("Cannot connect to H2O server. Please check that H2O is running at ", h2o.getBaseURL(tmpConn))
     else if (ip == "localhost" || ip == "127.0.0.1") {
       cat("\nH2O is not running yet, starting it now...\n")
+      command <- .h2o.checkJava()
+      javaVersionOutput <- system2(command, "-version", stdout=TRUE, stderr=TRUE)
+      v <- stringr::str_replace(javaVersionOutput[1], "^java version \"(.*)\"$","\\1")
+
+      # NOTE for developers: keep the following heuristic in sync with java code - see file connection.R near line 1850
+      if (!any(grepl("^1\\.7\\.",v), grepl("^1\\.8\\.",v), grepl("^9\\.", grepl("^10\\.", v)))) {
```

Comment:
This is correct. However, I think it would be better to reuse the functionality from line 527 and move the logic there. You attempt to call `java -version` another time here. This additional call is in vain. They store the result of the call and then they run checks against it. Why wouldn't we reuse the same mechanism ?

``` plain
  jver <- tryCatch({system2(command, "-version", stdout = TRUE, stderr = TRUE)}, # RESULT STORED HERE !!!!
      error = function(err) {
        print(err)
        stop("You have a 32-bit version of Java. H2O works best with 64-bit Java.\n",
        "Please download the latest Java SE JDK 8 from the following URL:\n",
        "http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html")
      }
    )

  if(any(grepl("GNU libgcj", jver))) { # CHECKED AGAINST ALREADY FETCHED OUTPUT !!!
    stop("Sorry, GNU Java is not supported for H2O.\n",
         "Please download the latest Java SE JDK 8 from the following URL:\n",
         "http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html")
  }
```

Reply:
true, I will change it this way